portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -23,13 +28,36 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content and handles data. */
+/**
+ * Servlet that handles input comments data and returns the contents with the limit number of
+ * maximum comments' number sent by viewers.
+ */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private List<String> commentsStorage = new ArrayList<>();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment");
+    PreparedQuery results = datastore.prepare(query);
+    List<String> commentsStorage = new ArrayList<>();
+    int maxCommentsNum = 3;
+    for (Entity entity : results.asIterable()) {
+      if (entity.getProperty("bound") != null) {
+        long boundLong = (long) entity.getProperty("bound");
+        int bound = Math.toIntExact(boundLong);
+        if (bound > maxCommentsNum) {
+          maxCommentsNum = bound;
+        }
+      }
+      String contents = (String) entity.getProperty("contents");
+      if (contents != "") {
+        commentsStorage.add(contents);
+      }
```

Comment:
This looks like for every comment we also store a bound, and we only use the largest bound in the datastore when filtering the list of comments? Is this the intended behaviour?

E.g. if we had in our datastore:
/* comment, bound */
{"comment 1", "2"},
{"comment 2", "3"},
{"comment 3", "2"},
{"comment 4", "1"}

then we would show the first 3 comments, right? And there is no way to show _less_ than 3 comments without deleting comment 2 and its bound?

Reply:
Yes... Now the idea is the maximum number of comments will be controlled by the maximum inserted number by all viewers.
I previously wanted to get the specific user's ID and grab the inserted number of that user. And meanwhile, all data will be stored in datastore rather than locally (which means directly as a variable for the class) to avoid multiple viewers commenting on the website at the same time and making some conflicts. However, I am blocked a while for the technical realisation for this idea. 
I will consider again for the idea.

Reply:
Yep ok, fair enough. Happy for you to leave it as is for now.

In the future if you do want to store this parameter in datastore, I would suggest having a separate table for storing it (and probably a separate form for updating it).

Another way to do it would be to add a `num-comments` parameter to the GET request.

A benefit of either of these approaches is that you wouldn't be storing the number of comments in the same table as the actual comments, and so you wouldn't have to load _every_ comment in the database for every GET request. Instead, you could use the [setLimit() method](https://cloud.google.com/datastore/docs/concepts/queries#cursors_limits_and_offsets) to only retrieve the comments you want.

Reply:
Yes! Thank you very much for your advice!
I think to have a separate table for storage and thus to limit the searching number is doable. 
Adding a num-comments parameter to the GET request is the preferred way by myself as well. I think if after adding the logging information, and I find there is a way that could make this process clearer, I will come back on this task.